### PR TITLE
DOC Improve documentation for LayerNorm

### LINKED ISSUE #51455: [docs] Improve documentation for LayerNorm
https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html: `The mean and standard-deviation are calculated separately over the last certain number dimensions which have to be of the shape specified by normalized_shape`

What's not clear from this excerpt (and the rest of the text):

1. What dimensions is the mean computed over? What is the shape of the computed mean/std? If it's the last few dimensions as suggested by this text, then does it mean that for BCHW layout, height and width must be fixed and passed at LayerNorm construction time? This seems wrong. Also expression "calculated separately over the last certain number of dims" may imply exactly the reverse: that the first dims are aggregated over. What seems to be happening is that "mean and standard-deviation are calculated over all except the first dimension (assumed batch dimension) and the specified last dimension (assumed channels dimensions)".
2. If the last few dimensions (including spatial/temporal) are aggregated over, why does LayerNorm need their (potentially dynamic) shape at construction time?

2. Is there any layout are assumed for a typical usecase (with fixed number of channels and various time/spatial dimension)? BTC? BCT? This is important, because PyTorch usually uses BC... logical layouts in other contexts

3. Could it include a visual explanation of what dimensions are aggregated over?

The examples given in the docs don't include dimension semantics (such as Channels, Time, Height, Width), so I couldn't extract any information from them.

E.g. an example of type:
```python
import torch

ln = torch.nn.LayerNorm(256, elementwise_affine = False)

# BTC
a = torch.randn(8, 1000, 256) * 100 + 200
print(a[0, ...].mean(), a[0, ...].std())
#tensor(199.9173) tensor(99.8517)

print(a[0, 0].mean(), a[0, 0].std())
#tensor(206.8604) tensor(104.7011)

b = ln(a)
print(b[0, ...].mean(), b[0, ...].std())
# tensor(2.2389e-09) tensor(1.0000)
print(a[0, 0].mean(), a[0, 0].std())
#tensor(-2.6077e-08) tensor(1.0020)

# what's going on?
# Does LayerNorm standardize every (b, ...) slice as illustrated in GroupNorm paper's figure?
# Does LayerNorm standardizes every (b, t, :) slice (i.e. standardizes embedding vectos themselves - does it even make sense)?
```
is much clearer.

Phrasing "over the last certain number dimensions" seems very ambiguous, as the aggregation seems to happen actually over not specified dimensions.

cc @jlin27 @mruberry @albanD @jbschlosser

### PR DESCRIPTION
**Pull request by thomasjpfan**
Closes https://github.com/pytorch/pytorch/issues/51455

I think the current implementation is aggregating over the correct dimensions. The shape of `normalized_shape` is only used to determine the dimensions to aggregate over. The actual values of `normalized_shape` are used when `elementwise_affine=True` to initialize the weights and biases.

This PR updates the docstring to clarify how `normalized_shape` is used. Here is a short script comparing the implementations for tensorflow and pytorch:

```python
import torch
import torch.nn as nn

import tensorflow as tf
from tensorflow.keras.layers import LayerNormalization

rng = np.random.RandomState()
x = rng.randn(10, 20, 64, 64).astype(np.float32)
# slightly non-trival
x[:, :10, ...] = x[:, :10, ...] * 10 + 20
x[:, 10:, ...] = x[:, 10:, ...] * 30 - 100

# Tensorflow Layer norm
x_tf = tf.convert_to_tensor(x)
layer_norm_tf = LayerNormalization(axis=[-3, -2, -1], epsilon=1e-5)
output_tf = layer_norm_tf(x_tf)
output_tf_np = output_tf.numpy()

# PyTorch Layer norm
x_torch = torch.as_tensor(x)
layer_norm_torch = nn.LayerNorm([20, 64, 64], elementwise_affine=False)
output_torch = layer_norm_torch(x_torch)
output_torch_np = output_torch.detach().numpy()

# check tensorflow and pytorch
torch.testing.assert_allclose(output_tf_np, output_torch_np)

# manual comutation
manual_output = ((x_torch - x_torch.mean(dim=(-3, -2, -1), keepdims=True)) / 
                 (x_torch.var(dim=(-3, -2, -1), keepdims=True, unbiased=False) + 1e-5).sqrt())

torch.testing.assert_allclose(output_torch, manual_output)
```

To get to the layer normalization as shown here:

Screen Shot 2021-05-29 at 2 13 52 PM

One needs to pass in `normalized_shape` with shape `x.dim() - 1` with the size of the channels and all spatial dimensions.

### CODE REVIEW

**Comment by vadimkantorov** [comment]
it would be great to have this sort of illustration in the docs for all normalization methods

**Comment by jbschlosser** [comment]
> it would be great to have this sort of illustration and pseudocode in the docs for all normalization methods

Agreed, I'm a big fan of these visualizations.

File: torch/nn/modules/normalization.py [hunk 1]
```diff
@@ -90,9 +90,10 @@ class LayerNorm(Module):
     .. math::
         y = \frac{x - \mathrm{E}[x]}{ \sqrt{\mathrm{Var}[x] + \epsilon}} * \gamma + \beta
 
-    The mean and standard-deviation are calculated separately over the last
-    certain number dimensions which have to be of the shape specified by
-    :attr:`normalized_shape`.
+    The mean and standard-deviation are calculated over a number of last dimensions
+    specifies by the shape of :attr:`normalized_shape`. For example, if the shape of
+    :attr:`normalized_shape` is 2, then the last two dimensions are used to compute
```
Comment (jbschlosser) [inline torch/nn/modules/normalization.py:95]: nit: I think it's more correct to refer to the "dimension" of `normalized_shape` rather than its shape, as a shape being 2 isn't quite right. Maybe something like:

```
The mean and standard-deviation are calculated over the last `D` dimensions, where `D` is the dimension of :attr:`normalized_shape`.
For example, if :attr:`normalized_shape` is ``(3, 5)`` (a 2-dimensional shape), the mean and standard-deviation are computed over the last 2 dimensions of the input (i.e. ``input.mean((-2, -1))``).
```

File: torch/nn/modules/normalization.py [hunk 2]
```diff
@@ -133,12 +134,16 @@ class LayerNorm(Module):
         >>> m = nn.LayerNorm(input.size()[1:])
         >>> # Without Learnable Parameters
         >>> m = nn.LayerNorm(input.size()[1:], elementwise_affine=False)
-        >>> # Normalize over last two dimensions
-        >>> m = nn.LayerNorm([10, 10])
         >>> # Normalize over last dimension of size 10
         >>> m = nn.LayerNorm(10)
+        >>> # Normalize over last three dimensions
```
Comment (jbschlosser) [inline torch/nn/modules/normalization.py:139]: What do you think about a small tweak to this example to use the standard CV dimensions? Something like:

```python
N, C, H, W = 20, 5, 10, 10
input = torch.randn(N, C, H, W)
```

with this comment changed to:
`Normalize over the last three dimensions (i.e. the channel and spatial dimensions)`

Then it's more clear how to achieve the standard normalization shown in the `LayerNorm` block diagram. I like how you described in the PR description how to achieve this, and I think it'd be valuable to have in the docs somehow. Ooh is it possible to add the image as well?

**Review by jbschlosser — COMMENTED** [review]
Thanks for tackling this! :)

**Comment by vadimkantorov** [comment]
For the vis to make sense and correspond to pseudocode, the example code should use semantic shape names. e.g. `torch.rand(b, h, w, c)`.

**Comment by vadimkantorov** [comment]
Can I somehow override `dependenciesFulfilled `? and await on it?

**Comment by thomasjpfan** [comment]
> maybe include also the typical NLP usecase

Do you think the typical use case is to normalize over the embedding dimension?

```python
batch, sentence_length, embedding_dim = 2, 4, 10
text_input = torch.randn(batch, sentence_length, embedding_dim)
layer_norm = nn.LayerNorm(embedding_dim)
layer_norm(text_input)
```

**Comment by vadimkantorov** [comment]
Yep, I think that's the typical usecase. What is the shape of mean/std statistics? is it `(embedding_dim, )`? (If so, then the visualization isn't correct, since different feature dims are not aggregated together)

**Comment by thomasjpfan** [comment]
> What is the shape of mean/std statistics?

mean's shape is: `(batch, sentence_length)`. Here is a small example:

```python
import torch
import torch.nn as nn
from torch.testing import assert_allclose

batch, sentence_length, embedding = 2, 3, 4
input = torch.arange(24, dtype=torch.float32).reshape(batch, sentence_length, embedding)
input
#tensor([[[ 0.,  1.,  2.,  3.],
#         [ 4.,  5.,  6.,  7.],
#         [ 8.,  9., 10., 11.]],

#        [[12., 13., 14., 15.],
#         [16., 17., 18., 19.],
#         [20., 21., 22., 23.]]])

layer_out = nn.LayerNorm(embedding)
output = layer_out(input)

#tensor([[[-1.3416, -0.4472,  0.4472,  1.3416],
#         [-1.3416, -0.4472,  0.4472,  1.3416],
#         [-1.3416, -0.4472,  0.4472,  1.3416]],

#        [[-1.3416, -0.4472,  0.4472,  1.3416],
#         [-1.3416, -0.4472,  0.4472,  1.3416],
#         [-1.3416, -0.4472,  0.4472,  1.3416]]],
#       grad_fn=<NativeLayerNormBackward>)

# manual
kwargs = {"dim": -1, "keepdim": True}
input_mean = input.mean(**kwargs)
num = input - input_mean
demon = (input.var(**kwargs, unbiased=False) + 1e-5).sqrt()
manual_output = num / demon

# same as manual
assert_allclose(output, manual_output)

input_mean.shape
# torch.Size([2, 3, 1])
```

(The 1 at the end is from `keepdim=True` to make broadcasting work nicely)

**Review by jbschlosser — APPROVED** [review]
LGTM!

**Comment by vadimkantorov** [comment]
For image, mabye it'd be good to have it as svg, then it would be easy to adapt it for other normalization ops

**Comment by vadimkantorov** [comment]
> mean's shape is: `(batch, sentence_length)`. Here is a small example:

If that's the case, the image seems incorrect - it should be a stripe corresponding to a single selected N AND a single selected H, W (i.e. T)

**Comment by thomasjpfan** [comment]
The image is specifically for the image use-case which computes statistics over `dim=(C, H, W)` and  the shaded blue region is normalized. (The H and W are collapsed in the representation)
 
For the NLP case, the statistics is over `dim=(embedding_dim)`, which results in this representation:

layer_norm_nlp

**Comment by vadimkantorov** [comment]
Also, this cube has transposed sides compared to the one in original message...

**Comment by thomasjpfan** [comment]
I am translating the cube from the image content to the NLP context as follows:

- `N` <-> Batch
- `C` <-> Sentence length
- `H, W` <-> Embedding dimension

This is consistent with how we construct the tensors: image -> `(N, C, H, W)`, and NLP -> `(batch, sentence_length, embedding_dim)`. Am I missing something here?

**Comment by vadimkantorov** [comment]
I think, it should be N <-> Batch; C <-> Embedding Dimension; H, W <-> Time Dimension

**Comment by vadimkantorov** [comment]
So I think normally the normalized_shape is C/embedding dim. And in vision for that you should movedim(-3, -1).

Also, it would be good if docs specified shapes of learned affine parameters

**Comment by thomasjpfan** [comment]
> Also, it would be good if docs specified shapes of learned affine parameters

Agreed.

> So I think normally the normalized_shape is C/embedding dim. And in vision for that you should movedim(-3, -1).

For the vision use-case, on page 3 of the [Group Norm](https://arxiv.org/abs/1803.08494) paper, near Eq (4) it states:

> meaning that LN computes μ and σ  along the (C,H,W) axes for each sample

The computation is always over the last three dimensions, which means swaping -3 and -1 shouldn't change the result.

> I think, it should be N <-> Batch; C <-> Embedding Dimension; H, W <-> Time Dimension

With this mental model, one needs to transpose the image to make sense for NLP. C is usually not the final dimension for vision (anymore) and embedding dimension is usually the final dimension. For example in hugging face's [SqueezeBertEmbeddings implementation](https://github.com/huggingface/transformers/blob/f8bd8c6c7eef0a7280cd58f89016ede5b77f142f/src/transformers/models/squeezebert/modeling_squeezebert.py#L86-L90) `LayerNorm` takes the output of `nn.Embedding` which places the embedding at the end.

**Comment by vadimkantorov** [comment]
That's the thing about GroupNorm paper, I think it describes LayerNomr not like it typically is used in NLP
